crac-plugin/src/main/java/io/micronaut/gradle/crac/MicronautCRaCPlugin.java

```diff
@@ -0,0 +1,247 @@
+package io.micronaut.gradle.crac;
+
+import com.bmuschko.gradle.docker.tasks.container.*;
+import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage;
+import com.bmuschko.gradle.docker.tasks.image.DockerPushImage;
+import com.bmuschko.gradle.docker.tasks.image.Dockerfile;
+import io.micronaut.gradle.MicronautExtension;
+import io.micronaut.gradle.MicronautRuntime;
+import io.micronaut.gradle.PluginsHelper;
+import io.micronaut.gradle.crac.tasks.CheckpointScriptTask;
+import io.micronaut.gradle.docker.DockerBuildStrategy;
+import io.micronaut.gradle.docker.model.MicronautDockerImage;
+import io.micronaut.gradle.docker.tasks.BuildLayersTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.plugins.ExtensionContainer;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.TaskContainer;
+import org.gradle.api.tasks.TaskProvider;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static io.micronaut.gradle.Strings.capitalize;
+
+public class MicronautCRaCPlugin implements Plugin<Project> {
+
+    public static final String CRAC_DEFAULT_BASE_IMAGE = "ubuntu:18.04";
+    public static final String CRAC_DEFAULT_BASE_IMAGE_PLATFORM = "linux/amd64";
+    private static final String CRAC_TASK_GROUP = "CRaC";
+    public static final String CRAC_CHECKPOINT = "crac_checkpoint";
+    public static final String BUILD_DOCKER_DIRECTORY = "docker/";
+
+    @Override
+    public void apply(Project project) {
+        configurePlugin(project);
+    }
+
+    private void configurePlugin(Project project) {
+        ExtensionContainer extensions = project.getExtensions();
+        MicronautExtension micronautExtension = extensions.getByType(MicronautExtension.class);
```

Comment:
:x: This will fail if the CrAC plugin is applied before, say, the `io.micronaut.application` plugin. If you need the `MicronautExtension` to be present, then you need to apply the `MicronautBasePlugin`. I guess this won't work either if you don't apply the `MicronautDockerPlugin`?

Reply:
Yeah, I couldn't decide on the best policy...or understand the inner workings...

Should this plugin just be additive to the application plugin?  ie:

```
plugins {
  id('io.micronaut.application')
  id('io.micronaut.crac')
}
```

Or should it apply the docker plugin and base plugins itself (as it needs them?)

Would that then fail if someone had:

```
plugins {
  id('io.micronaut.application')
  id('io.micronaut.docker')
  id('io.micronaut.crac')
}
```
?

Reply:
That's a good question. I'm leaning towards thinking that the plugin _adds a functionality_ to other plugins, similarly to the `aot` or `test-resources` plugin. Therefore, the plugin should, IMO, _react_ to the presence of the `docker` plugin. That is to say you would have:

```
plugins {
  id('io.micronaut.application')
  id('io.micronaut.crac')
}
```

(because `application` applies `docker`). Nevertheless, your plugin requires the Micronaut extension, so must apply the `MicronautBasePlugin`. Then it would be a matter of wrapping the configuration into a:

```
plugins.withPlugin("io.micronaut.docker") { ... }
```

block

Reply:
Cool.  Do we need to guess mistakes the user is going to make?

ie:

```
        PluginManager pluginManager = project.getPluginManager();
        pluginManager.apply(MicronautBasePlugin.class);
        if (!pluginManager.hasPlugin("io.micronaut.docker")) {
            throw new GradleException("The Micronaut CRaC plugin requires the Micronaut Docker plugin to be enabled");
        }
```

Reply:
No, you must not use `hasPlugin` (this method is eager, so you don't want to do this). You should use `pluginManager.withPlugin(...)` instead, which will execute _only_ when the docker plugin is applied.

Reply:
Urgh...  I wish Greedy methods were annotated as such

Cool, how's https://github.com/micronaut-projects/micronaut-gradle-plugin/pull/524/commits/14766a6a4572b6968dfd28a6ba915e5ba3b4b5c6 look?